### PR TITLE
Implements --real-client-ip-header option.

### LINKED ISSUE #424: X-Real-IP only meaningful to mostly CloudFlare users
## Expected Behavior

* Address displayed by logging should by default not include any non-standard semantics like treating `X-Real-IP` as the original forwarded IP.
* Most users with a reverse proxy will be using the `X-Forwarded-IP` as the defacto-standard for reflecting the original client IP.
* Using `X-Real-IP` is mostly done by CloudFlare, and gives a private organisation an unfair advantage for an otherwise neutral opensource project.

## Current Behavior

* The logger displays the address in `X-Real-IP` in various contexts.

## Possible Solution

* Add a configuration option to select what header to grab the real client IP from.

## Context

* Relates to desired other feature needed.

### PR DESCRIPTION
**Pull request by Izzette**
## Description

* The -real-client-ip-header determines what HTTP header is used for
  determining the "real client IP" of the remote client.
* The -real-client-ip-header option supports the following headers:
  X-Forwarded-For X-ProxyUser-IP and X-Real-IP (default).
* Introduces new realClientIPParser interface to allow for multiple
  polymorphic classes to decide how to determine the real client IP.
* TODO: implement the more standard, but more complex `Forwarded` HTTP
  header.

## Motivation and Context

* Support other headers besides `X-Real-IP` for logging.
* Support a future feature to allow whitelisted IPs behind a reverse-proxy to bypass authentication.

## How Has This Been Tested?

* Wrote new standard go tests to test the features implemented.
* TODO: write tests to confirm logging backwards compatibility.
  * Would love some advice on how capturing logs could be implemented.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

Fixes #424

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,7 @@
 
 ## Changes since v5.1.0
 
+- [#TBD](https://github.com/oauth2-proxy/oauth2-proxy/pull/TBD) Implements -real-client-ip-header option to select the header from which to obtain a proxied client's IP (@Izzette)
```
Comment (Izzette) [inline CHANGELOG.md:22]:
```suggestion
- [#503](https://github.com/oauth2-proxy/oauth2-proxy/pull/503) Implements -real-client-ip-header option to select the header from which to obtain a proxied client's IP (@Izzette)
```

File: pkg/logger/logger.go [hunk 2]
```diff
@@ -88,7 +90,7 @@ type Logger struct {
 	stdEnabled     bool
 	authEnabled    bool
 	reqEnabled     bool
-	reverseProxy   bool
+	getClientFunc  GetClientFunc
```
Comment (JoelSpeed) [inline pkg/logger/logger.go:93]: Wondering if this should just be `getClient`, I'm 50/50

File: realclientip.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		return nil, fmt.Errorf("unable to parse address (%s) from http.RemoteAddr", req.RemoteAddr)
+	} else if ip := net.ParseIP(ipStr); ip != nil {
+		return ip, nil
+	} else {
+		return nil, fmt.Errorf("unable to parse ip (%s)", ipStr)
+	}
+}
+
+// getClientString obtains the human readable string of the remote IP and optionally the real client IP if available
+func getClientString(p realClientIPParser, req *http.Request, full bool) (s string) {
```
Comment (JoelSpeed) [inline realclientip.go:75]: Are there tests for this? If there are, are there any tests for p being nil? I don't think I've seen any

File: docs/configuration/configuration.md [hunk 4]
```diff
@@ -100,6 +100,7 @@ An example [oauth2-proxy.cfg]({{ site.gitweb }}/contrib/oauth2-proxy.cfg.example
 | `-request-logging-format` | string | Template for request log lines | see [Logging Configuration](#logging-configuration) |
 | `-resource` | string | The resource that is protected (Azure AD only) | |
 | `-reverse-proxy` | bool | are we running behind a reverse proxy, controls whether headers like X-Real-Ip are accepted | false |
+| `-real-client-ip-header` | string | Header used to determine the real IP of the client (one of: X-Forwarded-For, X-Real-IP, or X-ProxyUser-IP) | X-Real-IP |
```
Comment (JoelSpeed) [inline docs/configuration/configuration.md:103]: Might be worth noting this relies on `ReverseProxy` being set, WDYT?

File: realclientip.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+// format "[<ip>]:<port>".  With-port and without-port formats are seamlessly supported concurrently.
+func (p xForwardedForClientIPParser) GetRealClientIP(h http.Header) (net.IP, error) {
+	var ipStr string
+	if realIP := h.Get(p.header); realIP != "" {
+		ipStr = realIP
+	} else {
+		return nil, nil
+	}
+
+	if commaIndex := strings.IndexRune(ipStr, ','); commaIndex != -1 {
```
Comment (JoelSpeed) [inline realclientip.go:46]: It's not apparent to me why this is being done, could you add a comment above this to explain it in a bit more details please?

File: pkg/logger/logger.go (hunk 2 above)
Comment (Izzette) [reply pkg/logger/logger.go:93]: I'm 50/50 too.  Since it's not too long, I think it's better to be explicit?

File: realclientip.go (hunk 3 above)
Comment (Izzette) [reply realclientip.go:75]: No, there's not.  I'll add some.

File: docs/configuration/configuration.md (hunk 4 above)
Comment (Izzette) [reply docs/configuration/configuration.md:103]: Definitely, I'll update it.

File: realclientip.go (hunk 5 above)
Comment (Izzette) [reply realclientip.go:46]:
```suggestion
	// Each successive proxy may append itself, comma separated, to the end of the X-Forwarded-for header.
	// Select only the first IP listed, as it it is the original client IP.
	if commaIndex := strings.IndexRune(ipStr, ','); commaIndex != -1 {
```

File: docs/configuration/configuration.md (hunk 4 above)
Comment (Izzette) [reply docs/configuration/configuration.md:103]:
```suggestion
| `-real-client-ip-header` | string | Header used to determine the real IP of the client, has no effect without the `-reverse-proxy` (one of: X-Forwarded-For, X-Real-IP, or X-ProxyUser-IP) | X-Real-IP |
```

File: docs/configuration/configuration.md [hunk 6]
```diff
@@ -100,6 +100,7 @@ An example [oauth2-proxy.cfg]({{ site.gitweb }}/contrib/oauth2-proxy.cfg.example
 | `-request-logging-format` | string | Template for request log lines | see [Logging Configuration](#logging-configuration) |
 | `-resource` | string | The resource that is protected (Azure AD only) | |
 | `-reverse-proxy` | bool | are we running behind a reverse proxy, controls whether headers like X-Real-Ip are accepted | false |
+| `-real-client-ip-header` | string | Header used to determine the real IP of the client, has no effect without the `-reverse-proxy` (one of: X-Forwarded-For, X-Real-IP, or X-ProxyUser-IP) | X-Real-IP |
```
Comment (JoelSpeed) [inline docs/configuration/configuration.md:103]:
```suggestion
| `-real-client-ip-header` | string | Header used to determine the real IP of the client, requires `-reverse-proxy` to be set (one of: X-Forwarded-For, X-Real-IP, or X-ProxyUser-IP) | X-Real-IP |
```

File: realclientip.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+func (p xForwardedForClientIPParser) GetRealClientIP(h http.Header) (net.IP, error) {
+	var ipStr string
+	if realIP := h.Get(p.header); realIP != "" {
+		ipStr = realIP
+	} else {
+		return nil, nil
+	}
+
+	// Each successive proxy may append itself, comma separated, to the end of the X-Forwarded-for header.
+	// Select only the first IP listed, as it it is the client IP recorded by the first proxy.
```
Comment (JoelSpeed) [inline realclientip.go:47]:
```suggestion
	// Select only the first IP listed, as it is the client IP recorded by the first proxy.
```

File: pkg/logger/logger.go (hunk 2 above)
Comment (JoelSpeed) [reply pkg/logger/logger.go:93]: @steakunderscore how do you feel?

File: docs/configuration/configuration.md (hunk 6 above)
Comment (Izzette) [reply docs/configuration/configuration.md:103]: I do feel like this implies an error will be produced if this option is specified without `-reverse-proxy`—which is not true—but maybe it should be?

File: realclientip.go (hunk 7 above)
Comment (Izzette) [reply realclientip.go:47]: Oops good catch.

File: docs/configuration/configuration.md (hunk 6 above)
Comment (JoelSpeed) [reply docs/configuration/configuration.md:103]: That's a fantastic idea! Could you look into updating the option validation to error if this is set but reverse proxy isn't? You would need to look at the flag to see if the value was changed, maybe that's not so easy

If it looks nasty then maybe we don't proceed with that, but I'm happy with how the phrasing I've suggested either way, unless one of the other maintainers disagrees cc @steakunderscore

Comment (Izzette) [reply docs/configuration/configuration.md:103]: Hmm, I'm not sure there's a not-totally-hacky way to determine whether or not -real-client-ip-header has been specified without significant code-change.  Additionally, I'm not sure it does make sense to crash if it's specified without `-reverse-proxy`, as it breaks the source-agnostic nature of the configuration: the behaviour changes whether it's set to `X-Real-IP` on the command line or from the default value.

I'll just update the language, and if @steakunderscore thinks we should change the behaviour here I can look into it more.

File: CHANGELOG.md [hunk 8]
```diff
@@ -19,6 +19,7 @@
 
 ## Changes since v5.1.0
 
+- [#503](https://github.com/oauth2-proxy/oauth2-proxy/pull/503) Implements -real-client-ip-header option to select the header from which to obtain a proxied client's IP (@Izzette)
```
Comment (JoelSpeed) [inline CHANGELOG.md:22]: Nit: Can we make the flag use a double dash as we are requiring this in the project from now on?
```suggestion
- [#503](https://github.com/oauth2-proxy/oauth2-proxy/pull/503) Implements --real-client-ip-header option to select the header from which to obtain a proxied client's IP (@Izzette)
```

**Comment by RichiCoder1** [comment]
Jumping in a bit late, but if you're adding support for all these, might be worth adding [Forwarded](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded) support too since that's the most recent standard.

File: CHANGELOG.md (hunk 8 above)
Comment (Izzette) [reply CHANGELOG.md:22]: Oh really?  Yes of course.

**Comment by Izzette** [comment]
> Jumping in a bit late, but if you're adding support for all these, might be worth adding [Forwarded](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded) support too since that's the most recent standard.

It's already noted as a TODO in the initial description, but because the specification of the forwarded header is more complex and my implementation is type-polymorphic I'm choosing the leave this out of this PR.  It can be included in another PR after this one is merged.

File: CHANGELOG.md (hunk 8 above)
Comment (Izzette) [reply CHANGELOG.md]: Done.

File: pkg/logger/logger.go (hunk 2 above)
Comment (Izzette) [reply pkg/logger/logger.go:93]: @steakunderscore ping

Comment (steakunderscore) [reply pkg/logger/logger.go:93]: Normally I would be against stating the type in the variable name, but here I think it's worth it for the clarity it adds.

File: CHANGELOG.md [hunk 9]
```diff
@@ -22,12 +22,14 @@
 - [#487](https://github.com/oauth2-proxy/oauth2-proxy/pull/487) Switch flags to StringSlice instead of StringArray
   - Options that take multiple arguments now split strings on commas if present
   - Eg `--foo=a,b,c,d` would result in the values `a`, `b`, `c` and `d` instead of a single `a,b,c,d` value as before
+
```
Comment (Izzette) [inline CHANGELOG.md:25]:
```suggestion
```

File: contrib/oauth2-proxy_autocomplete.sh [hunk 10]
```diff
@@ -20,6 +20,10 @@ _oauth2_proxy() {
 			COMPREPLY=( $(compgen -W "google azure facebook github keycloak gitlab linkedin login.gov digitalocean" -- ${cur}) )
 			return 0
 			;;
+		-real-client-ip-header)
```
Comment (Izzette) [inline contrib/oauth2-proxy_autocomplete.sh:23]: I'm unfamiliar with`@` as a wildcard in switch statements:
```suggestion
		-@real-client-ip-header)
```

or should I just:
```suggestion
		--real-client-ip-header)
```

**Comment by Izzette** [comment]
@steakunderscore @JoelSpeed, are we all happy, should we merge this?

File: contrib/oauth2-proxy_autocomplete.sh (hunk 10 above)
Comment (JoelSpeed) [reply contrib/oauth2-proxy_autocomplete.sh:23]: I think the @ is like an optional, but we aren't allowing single - anymore, so I think double dash is the right way to go here

File: docs/configuration/configuration.md [hunk 11]
```diff
@@ -90,6 +90,7 @@ An example [oauth2-proxy.cfg]({{ site.gitweb }}/contrib/oauth2-proxy.cfg.example
 | `--proxy-prefix` | string | the url root path that this proxy should be nested under (e.g. /`<oauth2>/sign_in`) | `"/oauth2"` |
 | `--proxy-websockets` | bool | enables WebSocket proxying | true |
 | `--pubjwk-url` | string | JWK pubkey access endpoint: required by login.gov | |
+| `--real-client-ip-header` | string | Header used to determine the real IP of the client, requires `-reverse-proxy` to be set (one of: X-Forwarded-For, X-Real-IP, or X-ProxyUser-IP) | X-Real-IP |
```
Comment (JoelSpeed) [inline docs/configuration/configuration.md:93]:
```suggestion
| `--real-client-ip-header` | string | Header used to determine the real IP of the client, requires `--reverse-proxy` to be set (one of: X-Forwarded-For, X-Real-IP, or X-ProxyUser-IP) | X-Real-IP |
```

**Pull request merged**